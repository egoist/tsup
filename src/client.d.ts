declare module '*?raw' {
  const value: string
  export default value
}
