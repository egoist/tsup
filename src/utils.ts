import fs from 'fs'
import JoyCon from 'joycon'
import stripJsonComments from 'strip-json-comments'
import resovleFrom from 'resolve-from'

const joycon = new JoyCon()

joycon.addLoader({
  test: /\.json$/,
  async load(filepath) {
    const content = stripJsonComments(
      await fs.promises.readFile(filepath, 'utf8')
    )
    return JSON.parse(content)
  },
})

// No backslash in path
function slash(input: string) {
  return input.replace(/\\/g, '/')
}

export type External =
  | string
  | RegExp
  | ((id: string, parentId?: string) => boolean)

export function isExternal(
  externals: External | External[],
  id: string,
  parentId?: string
) {
  id = slash(id)

  if (!Array.isArray(externals)) {
    externals = [externals]
  }

  for (const external of externals) {
    if (
      typeof external === 'string' &&
      (id === external || id.includes(`/node_modules/${external}/`))
    ) {
      return true
    }
    if (external instanceof RegExp) {
      if (external.test(id)) {
        return true
      }
    }
    if (typeof external === 'function') {
      if (external(id, parentId)) {
        return true
      }
    }
  }

  return false
}

export function loadTsConfig(cwd: string) {
  return joycon.load(['tsconfig.build.json', 'tsconfig.json'], cwd)
}

export async function getDeps(cwd: string) {
  const data = await loadPkg(cwd)

  const deps = Array.from(new Set([
    ...Object.keys(data.dependencies || {}),
    ...Object.keys(data.peerDependencies || {})
  ]))

  return deps
}

export async function loadPkg(cwd: string) {
  const { data } = await joycon.load(['package.json'], cwd)
  return data || {}
}

export function getBabel(): null | typeof import('@babel/core') {
  const p = resovleFrom.silent(process.cwd(), '@babel/core')
  return p && require(p)
}
