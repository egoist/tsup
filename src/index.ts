import path from 'path'
import fs from 'fs'
import { Worker } from 'worker_threads'
import type { Buildable, MarkRequired } from 'ts-essentials'
import { removeFiles, debouncePromise, slash } from './utils'
import { loadTsupConfig } from './load'
import glob from 'globby'
import { loadTsConfig } from 'bundle-require'
import { handleError, PrettyError } from './errors'
import resolveFrom from 'resolve-from'
import { parseArgsStringToArgv } from 'string-argv'
import type { ChildProcess } from 'child_process'
import execa from 'execa'
import kill from 'tree-kill'
import { version } from '../package.json'
import { createLogger, setSilent } from './log'
import { DtsConfig, Format, Options } from './options'
import { runEsbuild } from './esbuild'
import { shebang } from './plugins/shebang'
import { cjsSplitting } from './plugins/cjs-splitting'
import { PluginContainer } from './plugin'
import { es5 } from './plugins/es5'
import { sizeReporter } from './plugins/size-reporter'

export type { Format, Options }

export type NormalizedOptions = Omit<
  MarkRequired<Options, 'entry' | 'format' | 'outDir'>,
  'dts'
> & {
  dts?: DtsConfig
  tsconfigResolvePaths: Record<string, string[]>
  tsconfigDecoratorMetadata?: boolean
}

export const defineConfig = (
  options:
    | Options
    | Options[]
    | ((
        /** The options derived from CLI flags */
        overrideOptions: Options
      ) => Options | Options[])
) => options

const killProcess = ({
  pid,
  signal = 'SIGTERM',
}: {
  pid: number
  signal?: string | number
}) =>
  new Promise<unknown>((resolve) => {
    kill(pid, signal, resolve)
  })

const normalizeOptions = async (
  logger: ReturnType<typeof createLogger>,
  optionsFromConfigFile: Options | undefined,
  optionsOverride: Options
) => {
  const _options = {
    ...optionsFromConfigFile,
    ...optionsOverride,
  }
  const options: Buildable<NormalizedOptions> = {
    target: 'node12',
    format: ['cjs'],
    outDir: 'dist',
    ..._options,
    dts:
      typeof _options.dts === 'boolean'
        ? _options.dts
          ? {}
          : undefined
        : typeof _options.dts === 'string'
        ? { entry: _options.dts }
        : _options.dts,
  }

  setSilent(options.silent)

  const entry = options.entry || options.entryPoints

  if (!entry || Object.keys(entry).length === 0) {
    throw new PrettyError(`No input files, try "tsup <your-file>" instead`)
  }

  if (Array.isArray(entry)) {
    options.entry = await glob(entry)
    // Ensure entry exists
    if (!options.entry || options.entry.length === 0) {
      throw new PrettyError(`Cannot find ${entry}`)
    } else {
      logger.info('CLI', `Building entry: ${options.entry.join(', ')}`)
    }
  } else {
    Object.keys(entry).forEach((alias) => {
      const filename = entry[alias]!
      if (!fs.existsSync(filename)) {
        throw new PrettyError(`Cannot find ${alias}: ${filename}`)
      }
    })
    logger.info('CLI', `Building entry: ${JSON.stringify(entry)}`)
  }

  const tsconfig = loadTsConfig(process.cwd(), options.tsconfig)
  if (tsconfig.path) {
    logger.info(
      'CLI',
      `Using tsconfig: ${path.relative(process.cwd(), tsconfig.path)}`
    )
    options.tsconfig = tsconfig.path
    options.tsconfigResolvePaths = tsconfig.data?.compilerOptions?.paths || {}
    options.tsconfigDecoratorMetadata =
      tsconfig.data?.compilerOptions?.emitDecoratorMetadata
  } else if (options.tsconfig) {
    throw new PrettyError(`Cannot find tsconfig: ${options.tsconfig}`)
  }

  return options as NormalizedOptions
}

export async function build(_options: Options) {
  const config =
    _options.config === false ? {} : await loadTsupConfig(process.cwd())

  const configData =
    typeof config.data === 'function' ? config.data(_options) : config.data

  await Promise.all(
    [...(Array.isArray(configData) ? configData : [configData])].map(
      async (item) => {
        const logger = createLogger(item?.name)
        const options = await normalizeOptions(logger, item, _options)

        logger.info('CLI', `tsup v${version}`)

        if (config.path) {
          logger.info('CLI', `Using tsup config: ${config.path}`)
        }

        if (options.watch) {
          logger.info('CLI', 'Running in watch mode')
        }

        if (options.dts) {
          const worker = new Worker(path.join(__dirname, './rollup.js'))
          worker.postMessage({
            configName: item?.name,
            options: {
              ...options, // functions cannot be cloned
              esbuildPlugins: undefined,
              esbuildOptions: undefined,
            },
          })
          worker.on('message', (data) => {
            if (data === 'error') {
              process.exitCode = 1
            } else if (data === 'success') {
              process.exitCode = 0
            }
          })
        }

        if (!options.dts?.only) {
          let existingOnSuccess: ChildProcess | undefined
          /** Files imported by the entry */
          const buildDependencies: Set<string> = new Set()

          const killPreviousProcess = async () => {
            if (existingOnSuccess) {
              await killProcess({
                pid: existingOnSuccess.pid,
              })
              existingOnSuccess = undefined
            }
          }

          const debouncedBuildAll = debouncePromise(
            () => {
              return buildAll()
            },
            100,
            handleError
          )

          const buildAll = async () => {
            const killPromise = killPreviousProcess()
            // Store previous build dependencies in case the build failed
            // So we can restore it
            const previousBuildDependencies = new Set(buildDependencies)
            buildDependencies.clear()

            if (options.clean) {
              const extraPatterns = Array.isArray(options.clean)
                ? options.clean
                : []
              await removeFiles(
                ['**/*', '!**/*.d.ts', ...extraPatterns],
                options.outDir
              )
              logger.info('CLI', 'Cleaning output folder')
            }

            const css: Map<string, string> = new Map()
            await Promise.all([
              ...options.format.map(async (format, index) => {
                const pluginContainer = new PluginContainer([
                  shebang(),
                  ...(options.plugins || []),
                  cjsSplitting(),
                  es5(),
                  sizeReporter(),
                ])
                await pluginContainer.buildStarted()
                await runEsbuild(options, {
                  pluginContainer,
                  format,
                  css: index === 0 || options.injectStyle ? css : undefined,
                  logger,
                  buildDependencies,
                }).catch((error) => {
                  previousBuildDependencies.forEach(v => buildDependencies.add(v))
                  throw error
                })
              }),
            ])
            await killPromise
            if (options.onSuccess) {
              const parts = parseArgsStringToArgv(options.onSuccess)
              const exec = parts[0]
              const args = parts.splice(1)
              existingOnSuccess = execa(exec, args, {
                stdio: 'inherit',
              })
            }
          }

          const startWatcher = async () => {
            if (!options.watch) return

            const { watch } = await import('chokidar')

            const customIgnores = options.ignoreWatch
              ? Array.isArray(options.ignoreWatch)
                ? options.ignoreWatch
                : [options.ignoreWatch]
              : []

            const ignored = [
              '**/{.git,node_modules}/**',
              options.outDir,
              ...customIgnores,
            ]

            const watchPaths =
              typeof options.watch === 'boolean'
                ? '.'
                : Array.isArray(options.watch)
                ? options.watch.filter(
                    (path): path is string => typeof path === 'string'
                  )
                : options.watch

            logger.info(
              'CLI',
              `Watching for changes in ${
                Array.isArray(watchPaths)
                  ? watchPaths.map((v) => '"' + v + '"').join(' | ')
                  : '"' + watchPaths + '"'
              }`
            )
            logger.info(
              'CLI',
              `Ignoring changes in ${ignored
                .map((v) => '"' + v + '"')
                .join(' | ')}`
            )

            const watcher = watch(watchPaths, {
              ignoreInitial: true,
              ignorePermissionErrors: true,
              ignored,
            })
            watcher.on('all', (type, file) => {
              file = slash(file)
              if (!buildDependencies.has(file)) return

              logger.info('CLI', `Change detected: ${type} ${file}`)
              debouncedBuildAll()
            })
          }

          logger.info('CLI', `Target: ${options.target}`)

          await buildAll()

          startWatcher()
        }
      }
    )
  )
}
