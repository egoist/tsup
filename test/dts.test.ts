import path from 'node:path'
import { expect, test } from 'vitest'
import { slash } from '../src/utils'
import { getTestName, run } from './utils'

test('not bundle `package/subpath` in dts (resolve)', async () => {
  const { getFileContent } = await run(
    getTestName(),
    {
      'package.json': `{ "dependencies": { "foo": "*" } }`,
      'input.ts': `export const stuff: import('foo/bar').Foobar = { foo: 'foo', bar: 'bar' };`,
      'node_modules/foo/bar.d.ts': `export type Foobar = { foo: 'foo', bar: 'bar' }`,
      'node_modules/foo/package.json': `{ "name": "foo", "version": "0.0.0" }`,
    },
    {
      flags: ['--dts', '--dts-resolve'],
    },
  )
  const content = await getFileContent('dist/input.d.ts')
  expect(content).toMatchSnapshot()
})

test('enable --dts-resolve for specific module', async () => {
  const { getFileContent } = await run(getTestName(), {
    'input.ts': `export * from 'vue'
      export type {MarkRequired} from 'foo'
      `,
    'node_modules/foo/index.d.ts': `
      export type MarkRequired<T, RK extends keyof T> = Exclude<T, RK> & Required<Pick<T, RK>>
      `,
    'node_modules/foo/package.json': `{ "name": "foo", "version": "0.0.0" }`,
    'tsup.config.ts': `
      export default {
        dts: {
          resolve: ['foo']
        },
      }
      `,
  })
  const content = await getFileContent('dist/input.d.ts')
  expect(content).toMatchSnapshot()
})

test(`custom tsconfig should pass to dts plugin`, async () => {
  const { outFiles } = await run(getTestName(), {
    'input.ts': `export const foo = { name: 'foo'}`,
    'tsconfig.json': `{
        "compilerOptions": {
          "baseUrl":".",
          "target": "esnext",
          "incremental": true
        }
      }`,
    'tsconfig.build.json': `{
        "compilerOptions": {
          "baseUrl":".",
          "target": "esnext"
        }
      }`,
    'tsup.config.ts': `
        export default {
          entry: ['src/input.ts'],
          format: 'esm',
          tsconfig: './tsconfig.build.json',
          dts: {
            only: true
          }
        }
      `,
  })
  expect(outFiles).toEqual(['input.d.mts'])
})

test('should emit a declaration file per format', async () => {
  const { outFiles } = await run(getTestName(), {
    'input.ts': `export default 'foo'`,
    'tsup.config.ts': `
        export default {
          entry: ['src/input.ts'],
          format: ['esm', 'cjs'],
          dts: true
        }`,
  })
  expect(outFiles).toEqual([
    'input.d.mts',
    'input.d.ts',
    'input.js',
    'input.mjs',
  ])
})

test('should emit a declaration file per format (type: module)', async () => {
  const { outFiles } = await run(getTestName(), {
    'input.ts': `export default 'foo'`,
    'package.json': `{
      "type": "module"
    }`,
    'tsup.config.ts': `
        export default {
          entry: ['src/input.ts'],
          format: ['esm', 'cjs'],
          dts: true
        }`,
  })
  expect(outFiles).toEqual([
    'input.cjs',
    'input.d.cts',
    'input.d.ts',
    'input.js',
  ])
})

test('should emit dts chunks per format', async () => {
  const { outFiles } = await run(
    getTestName(),
    {
      'src/input1.ts': `
          import type { InternalType } from './shared.js'

          export function getValue(value: InternalType) {
            return value;
          }
      `,
      'src/input2.ts': `
          import type { InternalType } from './shared.js'

          export function getValue(value: InternalType) {
            return value;
          }
      `,
      'src/shared.ts': `export type InternalType = 'foo'`,
      'tsup.config.ts': `
        export default {
          entry: ['./src/input1.ts', './src/input2.ts'],
          format: ['esm', 'cjs'],
          dts: true
        }`,
    },
    { entry: [] },
  )
  expect(outFiles).toEqual([
    'input1.d.mts',
    'input1.d.ts',
    'input1.js',
    'input1.mjs',
    'input2.d.mts',
    'input2.d.ts',
    'input2.js',
    'input2.mjs',
    'shared-jWa9aNVo.d.mts',
    'shared-jWa9aNVo.d.ts',
  ])
})

test('should emit dts chunks per format (type: module)', async () => {
  const { outFiles } = await run(
    getTestName(),
    {
      'src/input1.ts': `
          import type { InternalType } from './shared.js'

          export function getValue(value: InternalType) {
            return value;
          }
      `,
      'src/input2.ts': `
          import type { InternalType } from './shared.js'

          export function getValue(value: InternalType) {
            return value;
          }
      `,
      'src/shared.ts': `export type InternalType = 'foo'`,
      'tsup.config.ts': `
        export default {
          entry: ['./src/input1.ts', './src/input2.ts'],
          format: ['esm', 'cjs'],
          dts: true
        }`,
      'package.json': `{
          "type": "module"
        }`,
    },
    { entry: [] },
  )
  expect(outFiles).toEqual([
    'input1.cjs',
    'input1.d.cts',
    'input1.d.ts',
    'input1.js',
    'input2.cjs',
    'input2.d.cts',
    'input2.d.ts',
    'input2.js',
    'shared-jWa9aNVo.d.cts',
    'shared-jWa9aNVo.d.ts',
  ])
})

test('should emit declaration files with experimentalDts', async () => {
  const files = {
    'package.json': `
        {
          "name": "tsup-playground",
          "private": true,
          "version": "0.0.0",
          "main": "dist/index.js",
          "module": "dist/index.mjs",
          "types": "dist/index.d.ts",
          "exports": {
              ".": {
                  "types": "./dist/index.d.ts",
                  "require": "./dist/index.js",
                  "import": "./dist/index.mjs",
                  "default": "./dist/index.js"
              },
              "./client": {
                  "types": "./dist/my-lib-client.d.ts",
                  "require": "./dist/my-lib-client.js",
                  "import": "./dist/my-lib-client.mjs",
                  "default": "./dist/my-lib-client.js"
              },
              "./server": {
                  "types": "./dist/server/index.d.ts",
                  "require": "./dist/server/index.js",
                  "import": "./dist/server/index.mjs",
                  "default": "./dist/server/index.js"
              }
          }
        }
    `,
    'tsconfig.json': `
        {
          "compilerOptions": {
              "target": "ES2020",
              "skipLibCheck": true,
              "noEmit": true
          },
          "include": ["./src"]
        }
    `,
    'tsup.config.ts': `
        export default {
          name: 'tsup',
          target: 'es2022',
          format: [
            'esm',
            'cjs'
          ],
          entry: {
            index: './src/index.ts',
            'my-lib-client': './src/client.ts',
            'server/index': './src/server.ts',
          },
        }
    `,
    'src/shared.ts': `
        export function sharedFunction<T>(value: T): T | null {
          return value || null
        }

        type sharedType = {
          shared: boolean
        }

        export type { sharedType }
    `,
    'src/server.ts': `
        export * from './shared'

        /**
         * Comment for server render function
         */
        export function render(options: ServerRenderOptions): string {
          return JSON.stringify(options)
        }

        export interface ServerRenderOptions {
          /**
           * Comment for ServerRenderOptions.stream
           *
           * @public
           *
           * @my_custom_tag
           */
          stream: boolean
        }

        export const serverConstant = 1

        export { serverConstant as serverConstantAlias }

        export class ServerClass {};

        export default function serverDefaultExport(options: ServerRenderOptions): void {};

        // Export a third party module as a namespace
        import * as ServerThirdPartyNamespace from 'react-dom';
        export { ServerThirdPartyNamespace }

        // Export a third party module
        export * from 'react-dom/server';

    `,
    'src/client.ts': `
        export * from './shared'

        export function render(options: ClientRenderOptions): string {
          return JSON.stringify(options)
        }

        export interface ClientRenderOptions {
          document: boolean
        }
    `,
    'src/index.ts': `
        export * from './client'
        export * from './shared'

        export const VERSION = '0.0.0' as const
    `,
  }
  const { outFiles, getFileContent } = await run(getTestName(), files, {
    entry: [],
    flags: ['--experimental-dts'],
  })
  const snapshots: string[] = []
  await Promise.all(
    outFiles
      .filter((outFile) => outFile.includes('.d.'))
      .map(async (outFile) => {
        const filePath = path.join('dist', outFile)
        const content = await getFileContent(filePath)
        snapshots.push(
          [
            '',
            '/'.repeat(70),
            `// ${path.posix.normalize(slash(filePath))}`,
            '/'.repeat(70),
            '',
            content,
          ].join('\n'),
        )
      }),
  )
  expect(snapshots.sort().join('\n')).toMatchSnapshot()
})

test('should only include exported declarations with experimentalDts', async () => {
  const files = {
    'package.json': `{ "name": "tsup-playground", "private": true }`,
    'tsconfig.json': `{ "compilerOptions": { "skipLibCheck": true } }`,
    'tsup.config.ts': `
        export default {
          entry: ['./src/entry1.ts', './src/entry2.ts']
        }
    `,
    'src/shared.ts': `
        export const declare1 = 'declare1'
        export const declare2 = 'declare2'
    `,
    'src/entry1.ts': `
        export { declare1 } from './shared'
    `,
    'src/entry2.ts': `
        export { declare2 } from './shared'
    `,
  }
  const { getFileContent } = await run(getTestName(), files, {
    entry: [],
    flags: ['--experimental-dts'],
  })

  const entry1dts = await getFileContent('dist/entry1.d.ts')
  const entry2dts = await getFileContent('dist/entry2.d.ts')

  expect(entry1dts).toContain('declare1')
  expect(entry1dts).not.toContain('declare2')

  expect(entry2dts).toContain('declare2')
  expect(entry2dts).not.toContain('declare1')
})

test('.d.ts files should be cleaned when --clean and --experimental-dts are provided', async () => {
  const filesFoo = {
    'package.json': `{ "name": "tsup-playground", "private": true }`,
    'foo.ts': `export const foo = 1`,
    'tsconfig.json': JSON.stringify(
      {
        compilerOptions: { skipLibCheck: true },
      },
      null,
      2,
    ),
  }

  const filesFooBar = {
    ...filesFoo,
    'bar.ts': `export const bar = 2`,
  }

  // First run with both foo and bar
  const result1 = await run(getTestName(), filesFooBar, {
    entry: ['foo.ts', 'bar.ts'],
    flags: ['--experimental-dts'],
  })

  expect(result1.outFiles).toContain('foo.d.ts')
  expect(result1.outFiles).toContain('foo.js')
  expect(result1.outFiles).toContain('bar.d.ts')
  expect(result1.outFiles).toContain('bar.js')

  // Second run with only foo
  const result2 = await run(getTestName(), filesFoo, {
    entry: ['foo.ts'],
    flags: ['--experimental-dts'],
  })

  // When --clean is not provided, the previous bar.* files should still exist
  expect(result2.outFiles).toContain('foo.d.ts')
  expect(result2.outFiles).toContain('foo.js')
  expect(result2.outFiles).toContain('bar.d.ts')
  expect(result2.outFiles).toContain('bar.js')

  // Third run with only foo and --clean
  const result3 = await run(getTestName(), filesFoo, {
    entry: ['foo.ts'],
    flags: ['--experimental-dts', '--clean'],
  })

  // When --clean is provided, the previous bar.* files should be deleted
  expect(result3.outFiles).toContain('foo.d.ts')
  expect(result3.outFiles).toContain('foo.js')
  expect(result3.outFiles).not.toContain('bar.d.ts')
  expect(result3.outFiles).not.toContain('bar.js')
})

test('dts only: ignore files', async () => {
  const { outFiles } = await run(
    getTestName(),
    {
      'input.ts': `
      import './style.scss'

      export const a = 1
      `,
      'style.scss': `
      @keyframes gallery-loading-spinner {
        0% {}
      }
      `,
    },
    {
      entry: ['input.ts'],
      flags: ['--dts-only'],
    },
  )
  expect(outFiles).toMatchInlineSnapshot(`
    [
      "input.d.ts",
    ]
  `)
})

test('declaration files with multiple entrypoints #316', async () => {
  const { getFileContent } = await run(
    getTestName(),
    {
      'src/index.ts': `export const foo = 1`,
      'src/bar/index.ts': `export const bar = 'bar'`,
    },
    { flags: ['--dts'], entry: ['src/index.ts', 'src/bar/index.ts'] },
  )
  expect(
    await getFileContent('dist/index.d.ts'),
    'dist/index.d.ts',
  ).toMatchSnapshot()
  expect(
    await getFileContent('dist/bar/index.d.ts'),
    'dist/bar/index.d.ts',
  ).toMatchSnapshot()
})

test('custom dts output extension', async ({ expect, task }) => {
  const { outFiles } = await run(
    getTestName(),
    {
      'src/types.ts': `export type Person = { name: string }`,
      'src/index.ts': `export const foo = [1, 2, 3]\nexport type { Person } from './types'`,
      'tsup.config.ts': `export default {
        name: '${task.name}',
        entry: { index: 'src/index.ts' },
        dts: true,
        format: ['esm', 'cjs'],
        outExtension({ format }) {
          return {
            js: format === 'esm' ? '.cjs' : '.mjs',
            dts: format === 'esm' ? '.d.cts' : '.d.mts',
          }
        },
    }`,
      'package.json': JSON.stringify(
        {
          name: 'custom-dts-output-extension',
          description: task.name,
          type: 'module',
        },
        null,
        2,
      ),
      'tsconfig.json': JSON.stringify(
        {
          compilerOptions: {
            outDir: './dist',
            rootDir: './src',
            moduleResolution: 'Bundler',
            module: 'ESNext',
            strict: true,
            skipLibCheck: true,
          },
          include: ['src'],
        },
        null,
        2,
      ),
    },
    {
      entry: [],
    },
  )
  expect(outFiles).toStrictEqual([
    'index.cjs',
    'index.d.cts',
    'index.d.mts',
    'index.mjs',
  ])
})
