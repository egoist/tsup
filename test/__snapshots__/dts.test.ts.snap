// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`declaration files with multiple entrypoints #316 1`] = `
"declare const foo = 1;

export { foo };
"
`;

exports[`declaration files with multiple entrypoints #316 2`] = `
"declare const bar = "bar";

export { bar };
"
`;

exports[`enable --dts-resolve for specific module 1`] = `
"export * from 'vue';

type MarkRequired<T, RK extends keyof T> = Exclude<T, RK> & Required<Pick<T, RK>>

export type { MarkRequired };
"
`;

exports[`not bundle \`package/subpath\` in dts (resolve) 1`] = `
"import * as foo_bar from 'foo/bar';

declare const stuff: foo_bar.Foobar;

export { stuff };
"
`;

exports[`should emit declaration files with experimentalDts 1`] = `
"
//////////////////////////////////////////////////////////////////////
// dist/_tsup-dts-rollup.d.mts
//////////////////////////////////////////////////////////////////////

import { PipeableStream } from 'react-dom/server';
import { ReactDOMServerReadableStream } from 'react-dom/server';
import { renderToNodeStream } from 'react-dom/server';
import { renderToPipeableStream } from 'react-dom/server';
import { RenderToPipeableStreamOptions } from 'react-dom/server';
import { renderToReadableStream } from 'react-dom/server';
import { RenderToReadableStreamOptions } from 'react-dom/server';
import { renderToStaticMarkup } from 'react-dom/server';
import { renderToStaticNodeStream } from 'react-dom/server';
import { renderToString } from 'react-dom/server';
import { ServerOptions } from 'react-dom/server';
import * as ServerThirdPartyNamespace from 'react-dom';
import { version } from 'react-dom/server';

declare interface ClientRenderOptions {
    document: boolean;
}
export { ClientRenderOptions }
export { ClientRenderOptions as ClientRenderOptions_alias_1 }

export declare function default_alias(options: ServerRenderOptions): void;

export { PipeableStream }

export { ReactDOMServerReadableStream }

declare function render(options: ClientRenderOptions): string;
export { render }
export { render as render_alias_1 }

/**
 * Comment for server render function
 */
export declare function render_alias_2(options: ServerRenderOptions): string;

export { renderToNodeStream }

export { renderToPipeableStream }

export { RenderToPipeableStreamOptions }

export { renderToReadableStream }

export { RenderToReadableStreamOptions }

export { renderToStaticMarkup }

export { renderToStaticNodeStream }

export { renderToString }

export declare class ServerClass {
}

declare const serverConstant = 1;
export { serverConstant }
export { serverConstant as serverConstantAlias }

export { ServerOptions }

export declare interface ServerRenderOptions {
    /**
     * Comment for ServerRenderOptions.stream
     *
     * @public
     *
     * @my_custom_tag
     */
    stream: boolean;
}

export { ServerThirdPartyNamespace }

declare function sharedFunction<T>(value: T): T | null;
export { sharedFunction }
export { sharedFunction as sharedFunction_alias_1 }
export { sharedFunction as sharedFunction_alias_2 }
export { sharedFunction as sharedFunction_alias_3 }

declare type sharedType = {
    shared: boolean;
};
export { sharedType }
export { sharedType as sharedType_alias_1 }
export { sharedType as sharedType_alias_2 }
export { sharedType as sharedType_alias_3 }

export declare const VERSION: "0.0.0";

export { version }

export { }


//////////////////////////////////////////////////////////////////////
// dist/_tsup-dts-rollup.d.ts
//////////////////////////////////////////////////////////////////////

import { PipeableStream } from 'react-dom/server';
import { ReactDOMServerReadableStream } from 'react-dom/server';
import { renderToNodeStream } from 'react-dom/server';
import { renderToPipeableStream } from 'react-dom/server';
import { RenderToPipeableStreamOptions } from 'react-dom/server';
import { renderToReadableStream } from 'react-dom/server';
import { RenderToReadableStreamOptions } from 'react-dom/server';
import { renderToStaticMarkup } from 'react-dom/server';
import { renderToStaticNodeStream } from 'react-dom/server';
import { renderToString } from 'react-dom/server';
import { ServerOptions } from 'react-dom/server';
import * as ServerThirdPartyNamespace from 'react-dom';
import { version } from 'react-dom/server';

declare interface ClientRenderOptions {
    document: boolean;
}
export { ClientRenderOptions }
export { ClientRenderOptions as ClientRenderOptions_alias_1 }

export declare function default_alias(options: ServerRenderOptions): void;

export { PipeableStream }

export { ReactDOMServerReadableStream }

declare function render(options: ClientRenderOptions): string;
export { render }
export { render as render_alias_1 }

/**
 * Comment for server render function
 */
export declare function render_alias_2(options: ServerRenderOptions): string;

export { renderToNodeStream }

export { renderToPipeableStream }

export { RenderToPipeableStreamOptions }

export { renderToReadableStream }

export { RenderToReadableStreamOptions }

export { renderToStaticMarkup }

export { renderToStaticNodeStream }

export { renderToString }

export declare class ServerClass {
}

declare const serverConstant = 1;
export { serverConstant }
export { serverConstant as serverConstantAlias }

export { ServerOptions }

export declare interface ServerRenderOptions {
    /**
     * Comment for ServerRenderOptions.stream
     *
     * @public
     *
     * @my_custom_tag
     */
    stream: boolean;
}

export { ServerThirdPartyNamespace }

declare function sharedFunction<T>(value: T): T | null;
export { sharedFunction }
export { sharedFunction as sharedFunction_alias_1 }
export { sharedFunction as sharedFunction_alias_2 }
export { sharedFunction as sharedFunction_alias_3 }

declare type sharedType = {
    shared: boolean;
};
export { sharedType }
export { sharedType as sharedType_alias_1 }
export { sharedType as sharedType_alias_2 }
export { sharedType as sharedType_alias_3 }

export declare const VERSION: "0.0.0";

export { version }

export { }


//////////////////////////////////////////////////////////////////////
// dist/index.d.mts
//////////////////////////////////////////////////////////////////////

export { VERSION } from './_tsup-dts-rollup.mjs';
export { render_alias_1 as render } from './_tsup-dts-rollup.mjs';
export { ClientRenderOptions_alias_1 as ClientRenderOptions } from './_tsup-dts-rollup.mjs';
export { sharedFunction_alias_1 as sharedFunction } from './_tsup-dts-rollup.mjs';
export { sharedType_alias_1 as sharedType } from './_tsup-dts-rollup.mjs';


//////////////////////////////////////////////////////////////////////
// dist/index.d.ts
//////////////////////////////////////////////////////////////////////

export { VERSION } from './_tsup-dts-rollup.js';
export { render_alias_1 as render } from './_tsup-dts-rollup.js';
export { ClientRenderOptions_alias_1 as ClientRenderOptions } from './_tsup-dts-rollup.js';
export { sharedFunction_alias_1 as sharedFunction } from './_tsup-dts-rollup.js';
export { sharedType_alias_1 as sharedType } from './_tsup-dts-rollup.js';


//////////////////////////////////////////////////////////////////////
// dist/my-lib-client.d.mts
//////////////////////////////////////////////////////////////////////

export { render } from './_tsup-dts-rollup.mjs';
export { ClientRenderOptions } from './_tsup-dts-rollup.mjs';
export { sharedFunction } from './_tsup-dts-rollup.mjs';
export { sharedType } from './_tsup-dts-rollup.mjs';


//////////////////////////////////////////////////////////////////////
// dist/my-lib-client.d.ts
//////////////////////////////////////////////////////////////////////

export { render } from './_tsup-dts-rollup.js';
export { ClientRenderOptions } from './_tsup-dts-rollup.js';
export { sharedFunction } from './_tsup-dts-rollup.js';
export { sharedType } from './_tsup-dts-rollup.js';


//////////////////////////////////////////////////////////////////////
// dist/server/index.d.mts
//////////////////////////////////////////////////////////////////////

export { render_alias_2 as render } from '../_tsup-dts-rollup.mjs';
export { default_alias as default } from '../_tsup-dts-rollup.mjs';
export { ServerRenderOptions } from '../_tsup-dts-rollup.mjs';
export { serverConstant } from '../_tsup-dts-rollup.mjs';
export { serverConstantAlias } from '../_tsup-dts-rollup.mjs';
export { ServerClass } from '../_tsup-dts-rollup.mjs';
export { ServerThirdPartyNamespace } from '../_tsup-dts-rollup.mjs';
export { sharedFunction_alias_2 as sharedFunction } from '../_tsup-dts-rollup.mjs';
export { sharedType_alias_2 as sharedType } from '../_tsup-dts-rollup.mjs';
export { renderToPipeableStream } from '../_tsup-dts-rollup.mjs';
export { renderToString } from '../_tsup-dts-rollup.mjs';
export { renderToNodeStream } from '../_tsup-dts-rollup.mjs';
export { renderToStaticMarkup } from '../_tsup-dts-rollup.mjs';
export { renderToStaticNodeStream } from '../_tsup-dts-rollup.mjs';
export { renderToReadableStream } from '../_tsup-dts-rollup.mjs';
export { RenderToPipeableStreamOptions } from '../_tsup-dts-rollup.mjs';
export { PipeableStream } from '../_tsup-dts-rollup.mjs';
export { ServerOptions } from '../_tsup-dts-rollup.mjs';
export { RenderToReadableStreamOptions } from '../_tsup-dts-rollup.mjs';
export { ReactDOMServerReadableStream } from '../_tsup-dts-rollup.mjs';
export { version } from '../_tsup-dts-rollup.mjs';


//////////////////////////////////////////////////////////////////////
// dist/server/index.d.ts
//////////////////////////////////////////////////////////////////////

export { render_alias_2 as render } from '../_tsup-dts-rollup.js';
export { default_alias as default } from '../_tsup-dts-rollup.js';
export { ServerRenderOptions } from '../_tsup-dts-rollup.js';
export { serverConstant } from '../_tsup-dts-rollup.js';
export { serverConstantAlias } from '../_tsup-dts-rollup.js';
export { ServerClass } from '../_tsup-dts-rollup.js';
export { ServerThirdPartyNamespace } from '../_tsup-dts-rollup.js';
export { sharedFunction_alias_2 as sharedFunction } from '../_tsup-dts-rollup.js';
export { sharedType_alias_2 as sharedType } from '../_tsup-dts-rollup.js';
export { renderToPipeableStream } from '../_tsup-dts-rollup.js';
export { renderToString } from '../_tsup-dts-rollup.js';
export { renderToNodeStream } from '../_tsup-dts-rollup.js';
export { renderToStaticMarkup } from '../_tsup-dts-rollup.js';
export { renderToStaticNodeStream } from '../_tsup-dts-rollup.js';
export { renderToReadableStream } from '../_tsup-dts-rollup.js';
export { RenderToPipeableStreamOptions } from '../_tsup-dts-rollup.js';
export { PipeableStream } from '../_tsup-dts-rollup.js';
export { ServerOptions } from '../_tsup-dts-rollup.js';
export { RenderToReadableStreamOptions } from '../_tsup-dts-rollup.js';
export { ReactDOMServerReadableStream } from '../_tsup-dts-rollup.js';
export { version } from '../_tsup-dts-rollup.js';
"
`;
